src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/ApplicationAssembliesProvider.cs

```diff
@@ -0,0 +1,256 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc.Core;
+using Microsoft.Extensions.DependencyModel;
+
+namespace Microsoft.AspNetCore.Mvc.ApplicationParts
+{
+    internal class ApplicationAssembliesProvider
+    {
+        internal static HashSet<string> ReferenceAssemblies { get; } = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+        {
+            "Microsoft.AspNetCore.Mvc",
+            "Microsoft.AspNetCore.Mvc.Abstractions",
+            "Microsoft.AspNetCore.Mvc.ApiExplorer",
+            "Microsoft.AspNetCore.Mvc.Core",
+            "Microsoft.AspNetCore.Mvc.Cors",
+            "Microsoft.AspNetCore.Mvc.DataAnnotations",
+            "Microsoft.AspNetCore.Mvc.Formatters.Json",
+            "Microsoft.AspNetCore.Mvc.Formatters.Xml",
+            "Microsoft.AspNetCore.Mvc.Localization",
+            "Microsoft.AspNetCore.Mvc.Razor",
+            "Microsoft.AspNetCore.Mvc.Razor.Extensions",
+            "Microsoft.AspNetCore.Mvc.RazorPages",
+            "Microsoft.AspNetCore.Mvc.TagHelpers",
+            "Microsoft.AspNetCore.Mvc.ViewFeatures",
+        };
+
+        /// <summary>
+        /// Returns an ordered list of application assemblies.
+        /// <para>
+        /// The order is as follows:
+        /// * Entry assembly
+        /// * Assemblies specified in the application's deps file ordered by name.
+        /// <para>
+        /// Each assembly is immediately followed by assemblies specified by annotated <see cref="RelatedAssemblyAttribute"/> ordered by name.
+        /// </para>
+        /// </para>
+        /// </summary>
+        public IEnumerable<Assembly> ResolveAssemblies(Assembly entryAssembly)
```

Comment:
This resolution logic is incomplete. You need to account for when the related assemblies might have gotten into the dependency context graph (like when they come from a nuget package).

`RelatedAssemblyAttribute should always win` over anything else to produce a stable result. Assemblies need to be ordered in the following way.
* Resolve the main assembly and its referenced assemblies.
* Resolve the related assemblies for all candidates.
* Filter the list of candidates by removing those that show up as related assemblies.

Order the remaining assemblies in the following order:
* Entry assembly.
* Entry assembly related assemblies.
* For the rest of the candidates (in lexicographical order for example)
  * Candidate assembly
  * Related assemblies for candidate.

Assemblies + RelatedAssembly should never form a graph (and you should probably throw in that case). There are two ways a graph can be produced by this scheme:
* RelatedAssembly points to the parent assembly.
* Two different assemblies point to the same RelatedAssembly.

The emphasis should be on RelatedAssemblyAttribute being a mechanism for you to bring "private" dependencies and not to reference common assemblies outside of the dependency context. (Point being, it's meant to be for stuff precompiled into a "satellite assembly" like precompiled views.

Reply:
The ordering was already taken care of. We were missing checks for cyclical cases which I added as part of my most recent iteration.

Reply:
Where?

Reply:
My bad, I don't know how to read code.